### PR TITLE
Move prelude in top of Ace Editor

### LINKED ISSUE #215: Better control over visibility of prelude elements
This is a feature request.

Currently, if I understand correctly, the code in `prelude.ml` is visible to the student, whereas the code in `prepare.ml` is not. Furthermore, `prepare.ml` can refer to (unqualified) names defined in `prelude.ml`, but *not* the other way around. This is too inflexible; for instance, I have an example where I need a type definition to be in `prelude.ml` (because the rest of the prelude depends on it) yet I would like it to be invisible (because I want the student to view this type as abstract).

Could we have finer-grained control over visibility? E.g., elements in `prelude.ml` could optionally be marked `[@invisible]`. (Perhaps, for the sake of symmetry, elements in `prepare.ml` could be marked `[@visible]`.)

### PR DESCRIPTION
**Pull request by beobillly**
Fix #215 
This PR move the prelude from the bottom of the Exercise pane to the begining of the tab editor.
We can hide the toplevel code with a button, the tab editor automatically adapts itself to the size of the toplevel code.
If the toplevel code is too big, use of the Toplevel window on the right to exercise is recommended.

A view with the prelude
learn ocaml prelude show

A view without the prelude
Learn ocaml prelude hide

### CODE REVIEW

**Comment by yurug** [comment]
Did you check that the layout is fine on a smartphone?

**Comment by beobillly** [comment]
the last commit fixed the layout on the smartphone with one and two colunms.

this one with 2 columns and toplevel code
smartphone 2

this one with 1 column and toplevel code
smatphone 1

this one with 1 column and no toplevel code
smartphone 1 hide

this one with 2 columns and no toplevel code
smartphone 2 hide

**Comment by beobillly** [comment]
My bad, this pull request does not refer to issue #215 but issue #168 .

**Comment by AltGr** [comment]
This looks good, thanks!

**Comment by beobillly** [comment]
I found some graphical issues in my pr, I'm working on them.
Don't merge my pr before I solved it !
Also I must add a scroll-bar to the toplevel code.

**Comment by yurug** [comment]
Are you done fixing the last issue of this PR @beobillly ?

**Comment by beobillly** [comment]
It's finally done !
I add a scroll bar on the prelude toplevel code when the code is longer than ~ 8 lines
I fixed some ace graphical issues which appened when the prelude was shown.
I fixed some bugs which were occuring when the editor pane content and the toplevel code were using a scrollbar (the content pane editor wasn't resized correctly).
The prelude toplvel code is shown by default.
It works with one and two columns and with mobile interface !

with the prelude hidden
ocaml prelude hide

with the prelude shown
ocaml prelude show

with prelude shwon and the scrolling at the end 
ocaml prelude scrolling end

with prelude hidden on mobile
ocaml prelude mobile hide

with prelude shown on mobile
ocaml prelude mobile show

with prelude resized on mobile
ocaml prelude mobile scrollbar bottom

File: demo-repository/exercises/demo/prelude.ml [hunk 1]
```diff
@@ -1,2 +1,41 @@
 (* Some code is loaded in the toplevel before your code. *)
 let greetings = "Hello world!"
+let greetings = "Hello world!"
```
Comment (yurug) [inline demo-repository/exercises/demo/prelude.ml]: Please clean up!

Comment (beobillly) [reply demo-repository/exercises/demo/prelude.ml]: Done

Comment (yurug) [reply demo-repository/exercises/demo/prelude.ml]: There are still 2 lines instead of one.

Comment (yurug) [reply demo-repository/exercises/demo/prelude.ml]: Thanks

**Pull request merged**